plugins/MauticCrmBundle/Api/ConnectwiseApi.php

```diff
@@ -59,103 +66,162 @@ protected function request($endpoint, $parameters = [], $method = 'GET')
     }
 
     /**
+     * @param array $params
+     *
      * @return mixed|string
      *
      * @throws ApiErrorException
      */
-    public function createContact($params)
+    public function createContact(array $params)
     {
         return $this->request('company/contacts', $params, 'POST');
     }
 
     /**
+     * @param array $params
+     * @param       $id
+     *
      * @return mixed|string
      *
      * @throws ApiErrorException
      */
-    public function updateContact($params, $id)
+    public function updateContact(array $params, $id)
     {
         return $this->request('company/contacts/'.$id, $params, 'PATCH');
     }
 
     /**
-     * @return mixed|string
+     * @return array
      *
      * @throws ApiErrorException
      */
     public function getCampaigns()
     {
-        return $this->request('marketing/groups');
+        return $this->fetchAllRecords('marketing/groups');
     }
 
     /**
-     * @param $campaignId
-     * @param $params
+     * @param     $campaignId
+     * @param int $page
      *
      * @return mixed|string
      *
      * @throws ApiErrorException
      */
-    public function getCampaignMembers($campaignId)
+    public function getCampaignMembers($campaignId, $page = 1)
     {
-        return $this->request('marketing/groups/'.$campaignId.'/contacts');
+        return $this->request('marketing/groups/'.$campaignId.'/contacts', ['page' => $page]);
     }
 
     /**
      * https://{connectwiseSite}/v4_6_release/apis/3.0/sales/activities/types.
      *
-     * @param array $params
-     *
-     * @return mixed|string
+     * @return array
      *
      * @throws ApiErrorException
      */
-    public function getActivityTypes($params = [])
+    public function getActivityTypes()
     {
-        return $this->request('sales/activities/types');
+        return $this->fetchAllRecords('sales/activities/types');
     }
 
+    /**
+     * @param array $params
+     *
+     * @return array
+     *
+     * @throws ApiErrorException
+     */
     public function postActivity($params = [])
     {
         return $this->request('sales/activities', $params, 'POST');
     }
 
-    public function getMembers($params = [])
+    /**
+     * @return array
+     *
+     * @throws ApiErrorException
+     */
+    public function getMembers()
     {
-        return $this->request('system/members');
+        return $this->fetchAllRecords('system/members');
+    }
+
+    /**
+     * @param $endpoint
+     *
+     * @return array
+     *
+     * @throws ApiErrorException
+     */
+    public function fetchAllRecords($endpoint)
+    {
+        $page        = 1;
+        $pageSize    = ConnectwiseIntegration::PAGESIZE;
+        $allRecords  = [];
+        while ($pagedRecords = $this->request($endpoint, ['page' => $page, 'pageSize' => $pageSize])) {
+            $allRecords = array_merge($allRecords, $pagedRecords);
+            ++$page;
+
+            if (count($pagedRecords) < $pageSize) {
```

Comment:
If the total records is evenly divisible by the page size, this if won’t catch and a final request will be made that throws a 404 and cause the process to fail.

Reply:
Assuming this happened? I couldn’t find documentation that CW told us exactly how many pages it had so this was the alternative. Guess we’ll have to put catch in for the 404 to prevent this.

Reply:
Well, searching through the logs with the response headers recorded, they use a Link header with URLs. I'll see if I can tap into that instead of the record count comparison.

Reply:
Ugh; except we don't feed headers back anywhere which means more refactoring.

Reply:
I’d just do a try catch inside the while and if catch just break